aquadoggo/src/domain.rs

```diff
@@ -0,0 +1,1060 @@
+
+    // If no document_view_id is passed then this is a request for publishing a CREATE operation
+    // and we return the args for the next free log by this author.
+    if document_view_id.is_none() {
+        let log_id = next_log_id(store, public_key).await?;
+        next_args.log_id = log_id.into();
+        return Ok(next_args);
+    }
+
+    ///////////////////////////
+    // DETERMINE DOCUMENT ID //
+    ///////////////////////////
+
+    // We can unwrap here as we know document_view_id is some.
+    let document_view_id = document_view_id.unwrap();
+
+    // Get the document_id for this document_view_id. This performs several validation steps (check
+    // method doc string).
+    let document_id = get_validate_document_id_for_view_id(store, document_view_id).await?;
+
+    // Check the document is not deleted.
+    ensure_document_not_deleted(store, &document_id).await?;
+
+    /////////////////////////
+    // DETERMINE NEXT ARGS //
+    /////////////////////////
+
+    // Retrieve the log_id for the found document_id and author.
+    //
+    // @TODO: (lolz, this method is just called `get()`)
+    let log_id = store.get(public_key, &document_id).await?;
+
+    // Check if an existing log id was found for this author and document.
+    match log_id {
+        // If it wasn't found, we just calculate the next log id safely and return the next args.
+        None => {
+            let next_log_id = next_log_id(store, public_key).await?;
+            next_args.log_id = next_log_id.into()
+        }
+        // If one was found, we need to get the backlink and skiplink, and safely increment the seq num.
+        Some(log_id) => {
+            // Get the latest entry in this log.
+            let latest_entry = store.get_latest_entry(public_key, &log_id).await?;
+
+            // Determine the next sequence number by incrementing one from the latest entry seq num.
+            //
+            // If the latest entry is None, then we must be at seq num 1.
+            let seq_num = match latest_entry {
+                Some(ref latest_entry) => {
+                    let mut latest_seq_num = latest_entry.seq_num();
+                    increment_seq_num(&mut latest_seq_num)
+                }
+                None => Ok(SeqNum::default()),
+            }?;
+
+            // Check if skiplink is required and if it is get the entry and return it's hash
+            let skiplink = if is_lipmaa_required(seq_num.as_u64()) {
+                // Determine skiplink ("lipmaa"-link) entry in this log.
+                Some(get_expected_skiplink(store, public_key, &log_id, &seq_num).await?)
+            } else {
+                None
+            }
+            .map(|entry| entry.hash());
+
+            next_args.backlink = latest_entry.map(|entry| entry.hash().into());
+            next_args.skiplink = skiplink.map(|hash| hash.into());
+            next_args.seq_num = seq_num.into();
+            next_args.log_id = log_id.into();
```

Comment:
I'd rather this wasn't a gql type.

Reply:
Yeh, just in the domain module. Seeing as the scope of these methods is quite clearly defined now, and they are to be used behind a public API, it seems neat that they only deal with p2panda core types and the store. I see them more like utility methods, rather high-level ones, but still not married to any "outside" protocol.  

As well as this, `validation` and `domain` will be moved into `p2panda-rs` later where we don't have a concept of gql.

Reply:
I think thats a good idea, do you want to do this as part of this PR?